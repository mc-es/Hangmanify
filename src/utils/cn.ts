import type { ClassValue } from 'clsx';
import clsx from 'clsx';
import { twMerge } from 'tailwind-merge';

/**
 * ### Combines conditional Tailwind class names and resolves conflicts
 *
 * @param {...ClassValue[]} inputs - Class names, booleans, objects, or arrays
 * @returns {string} Merged and deduplicated class string
 * @example
 * cn('p-4', isDark && 'bg-black') // 'p-4 bg-black' if isDark is true, otherwise 'p-4'
 */
export const cn = (...inputs: ClassValue[]): string => twMerge(clsx(...inputs));
