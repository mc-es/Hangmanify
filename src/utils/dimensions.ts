import { Dimensions as NativeDimensions, PixelRatio } from 'react-native';

const { width: deviceWidth, height: deviceHeight } = NativeDimensions.get('window');

const guidelineBaseWidth = 375;
const guidelineBaseHeight = 812;
const BLEND_ZONE = 5;

const BREAKPOINTS = {
  '2xs': { size: 320, scale: 0.65 },
  'xs-2xs': { size: 340, scale: 0.7 },
  xs: { size: 360, scale: 0.75 },
  'xs-sm': { size: 375, scale: 0.75 },
  sm: { size: 390, scale: 0.85 },
  'sm-md': { size: 420, scale: 0.9 },
  md: { size: 480, scale: 1 },
  'md-lg': { size: 540, scale: 1.05 },
  lg: { size: 600, scale: 1.1 },
  'lg-xl': { size: 680, scale: 1.15 },
  xl: { size: 760, scale: 1.2 },
  'xl-2xl': { size: 800, scale: 1.3 },
  '2xl': { size: 840, scale: 1.4 },
} as const;

const easeIn = (t: number): number => t * t;
const easeOut = (t: number): number => t * (2 - t);

const getScale = (width: number): number => {
  const breakpoints = Object.values(BREAKPOINTS).sort((x, y) => x.size - y.size);

  if (width <= breakpoints[0].size) return breakpoints[0].scale;
  if (width >= breakpoints[breakpoints.length - 1].size)
    return breakpoints[breakpoints.length - 1].scale;

  for (let i = 0; i < breakpoints.length - 1; i++) {
    const current = breakpoints[i];
    const next = breakpoints[i + 1];

    if (width >= current.size && width < next.size) {
      const scaleRange = next.scale - current.scale;
      const widthRange = next.size - current.size;
      const delta = width - current.size;

      let finalProgress;

      if (delta < BLEND_ZONE) {
        const t = delta / BLEND_ZONE;
        finalProgress = easeIn(t) * (BLEND_ZONE / widthRange);
      } else if (delta > widthRange - BLEND_ZONE) {
        const t = (delta - (widthRange - BLEND_ZONE)) / BLEND_ZONE;
        finalProgress =
          (widthRange - BLEND_ZONE) / widthRange + easeOut(t) * (BLEND_ZONE / widthRange);
      } else finalProgress = delta / widthRange;

      return current.scale + scaleRange * finalProgress;
    }
  }

  return 1;
};

const scale = getScale(deviceWidth);

// horizontal & vertical scale
const hs = (size: number): number => (deviceWidth / guidelineBaseWidth) * size * scale;
const vs = (size: number): number => (deviceHeight / guidelineBaseHeight) * size;

// width & height percent
const wp = (percent: number): number => (deviceWidth * percent) / 100;
const hp = (percent: number): number => (deviceHeight * percent) / 100;

// moderate scale
const ms = (size: number, factor = 0.5): number => size + (hs(size) - size) * factor;

// font scale
const fs = (size: number): number => {
  const fontScale = PixelRatio.getFontScale();
  return Math.round(PixelRatio.roundToNearestPixel(hs(size) * fontScale));
};

export const Dimensions = { hs, vs, wp, hp, ms, fs };
