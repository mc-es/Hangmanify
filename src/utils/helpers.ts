import type { TranslateOptions } from 'i18n-js';

function getNestedValue<T>(obj: T, key: string): unknown {
  return key.split('.').reduce((acc: unknown, curr: string): unknown => {
    if (acc !== null && typeof acc === 'object' && curr in acc)
      return (acc as Record<string, unknown>)[curr];

    return null;
  }, obj as unknown);
}
function interpolate(template: string, options: TranslateOptions): string {
  return template.replace(/{{\s*(\w+)\s*}}/g, (_match, key) =>
    key in options ? String(options[key]) : ''
  );
}

export const helpers = { getNestedValue, interpolate };
