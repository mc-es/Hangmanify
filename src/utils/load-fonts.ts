import * as Font from 'expo-font';

import { APP_FONTS } from 'assets/fonts';

import { FONTS } from 'src/constants/styles';

export const LoadFonts = async (): Promise<void> => {
  try {
    await Font.loadAsync({
      [FONTS.Nunito.Bold]: APP_FONTS.nunito.bold,
      [FONTS.Nunito.Light]: APP_FONTS.nunito.light,
      [FONTS.Nunito.Medium]: APP_FONTS.nunito.medium,
      [FONTS.Nunito.Regular]: APP_FONTS.nunito.regular,
      [FONTS.Nunito.SemiBold]: APP_FONTS.nunito.semibold,
      [FONTS.Poppins.Bold]: APP_FONTS.poppins.bold,
      [FONTS.Poppins.Light]: APP_FONTS.poppins.light,
      [FONTS.Poppins.Medium]: APP_FONTS.poppins.medium,
      [FONTS.Poppins.Regular]: APP_FONTS.poppins.regular,
      [FONTS.Poppins.SemiBold]: APP_FONTS.poppins.semibold,
    });
  } catch (error) {
    console.error('Error loading fonts', error);
  }
};
