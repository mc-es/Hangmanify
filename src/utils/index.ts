import { cn } from './cn';
import { dimensions } from './dimensions';
import { helpers } from './helpers';

export { cn, dimensions, helpers };
