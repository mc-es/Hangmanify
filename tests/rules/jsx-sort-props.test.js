/**
 * @fileoverview Tests for jsx-sort-props rule.
 * @see https://github.com/mces58/Hangmanify/blob/master/guides/rules/jsx-sort-props.md
 */
import rule from 'plugins/rules/jsx-sort-props.mjs';

import { runTest } from './utils.js';

const validCases = [
  {
    name: 'Props in correct order',
    filename: 'src/components/CorrectOrder.tsx',
    code: `<Component a="1" b="2" />`,
  },
  {
    name: 'Single prop',
    filename: 'src/screens/SingleProp.tsx',
    code: `<Component a="1" />`,
  },
  {
    name: 'No props',
    filename: 'src/pages/NoProps.tsx',
    code: `<Component />`,
  },
  {
    name: 'Spread only',
    filename: 'shared/elements/SpreadOnly.tsx',
    code: `<Component {...props} />`,
  },
  {
    name: 'Spread followed by sorted props',
    filename: 'app/ui/SpreadSorted.tsx',
    code: `<Component {...props} a="1" b="2" />`,
  },
  {
    name: 'Sorted with spread between',
    filename: 'src/screens/SpreadMiddleSorted.tsx',
    code: `<Component a="1" {...props} b="2" />`,
  },
  {
    name: 'Similar starting letters, correct order',
    filename: 'src/components/SimilarStartCorrect.tsx',
    code: `<Component a="2" aa="3" />`,
  },
  {
    name: 'Similar long names in correct order',
    filename: 'shared/elements/LongNamesCorrect.tsx',
    code: `<Component app="1" apple="2" />`,
  },
];

const invalidCases = [
  {
    name: 'Props in wrong order',
    filename: 'src/pages/WrongOrder.tsx',
    code: `<Component b="2" a="1" />`,
    expectedErrors: 1,
  },
  {
    name: 'Spread followed by unsorted props',
    filename: 'src/screens/SpreadUnsorted.tsx',
    code: `<Component {...props} b="2" a="1" />`,
    expectedErrors: 1,
  },
  {
    name: 'Unsorted with spread between',
    filename: 'app/ui/SpreadMiddleUnsorted.tsx',
    code: `<Component b="1" {...props} a="2" />`,
    expectedErrors: 1,
  },
  {
    name: 'Similar starting letters, wrong order',
    filename: 'shared/elements/SimilarStartWrong.tsx',
    code: `<Component aa="1" a="2" />`,
    expectedErrors: 1,
  },
  {
    name: 'Similar long names, wrong order',
    filename: 'src/components/LongNamesWrong.tsx',
    code: `<Component apple="1" app="2" />`,
    expectedErrors: 1,
  },
  {
    name: 'Multiline props wrong order',
    filename: 'src/screens/MultilineWrong.tsx',
    code: `
      <Component
        z="1"
        a="2"
      />`,
    expectedErrors: 1,
  },
  {
    name: 'Multiple unsorted props',
    filename: 'src/pages/MultipleUnsorted.tsx',
    code: `<Component c="3" b="2" a="1" />`,
    expectedErrors: 2,
  },
  {
    name: 'Boolean shorthand props unsorted',
    filename: 'src/components/BooleanUnsorted.tsx',
    code: `<Component b a />`,
    expectedErrors: 1,
  },
  {
    name: 'Duplicate props in wrong order',
    filename: 'app/ui/DuplicateProps.tsx',
    code: `<Component b="1" a="2" b="3" />`,
    expectedErrors: 1,
  },
];

runTest('jsx-sort-props', rule, validCases, invalidCases, /should come before/);
