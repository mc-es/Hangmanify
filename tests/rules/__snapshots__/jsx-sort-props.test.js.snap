// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jsx-sort-props rule invalid cases Boolean shorthand props unsorted 1`] = `
[
  {
    "column": 14,
    "endColumn": 15,
    "endLine": 1,
    "line": 1,
    "message": "JSX props should be sorted in alphabetical order. 'a' should come before 'b'.",
    "messageId": "unsortedProps",
    "nodeType": "JSXAttribute",
    "ruleId": "@custom-typescript-eslint/jsx-sort-props",
    "severity": 2,
  },
]
`;

exports[`jsx-sort-props rule invalid cases Duplicate props in wrong order 1`] = `
[
  {
    "column": 18,
    "endColumn": 23,
    "endLine": 1,
    "line": 1,
    "message": "JSX props should be sorted in alphabetical order. 'a' should come before 'b'.",
    "messageId": "unsortedProps",
    "nodeType": "JSXAttribute",
    "ruleId": "@custom-typescript-eslint/jsx-sort-props",
    "severity": 2,
  },
]
`;

exports[`jsx-sort-props rule invalid cases Multiline props wrong order 1`] = `
[
  {
    "column": 9,
    "endColumn": 14,
    "endLine": 4,
    "line": 4,
    "message": "JSX props should be sorted in alphabetical order. 'a' should come before 'z'.",
    "messageId": "unsortedProps",
    "nodeType": "JSXAttribute",
    "ruleId": "@custom-typescript-eslint/jsx-sort-props",
    "severity": 2,
  },
]
`;

exports[`jsx-sort-props rule invalid cases Multiple unsorted props 1`] = `
[
  {
    "column": 18,
    "endColumn": 23,
    "endLine": 1,
    "line": 1,
    "message": "JSX props should be sorted in alphabetical order. 'b' should come before 'c'.",
    "messageId": "unsortedProps",
    "nodeType": "JSXAttribute",
    "ruleId": "@custom-typescript-eslint/jsx-sort-props",
    "severity": 2,
  },
  {
    "column": 24,
    "endColumn": 29,
    "endLine": 1,
    "line": 1,
    "message": "JSX props should be sorted in alphabetical order. 'a' should come before 'b'.",
    "messageId": "unsortedProps",
    "nodeType": "JSXAttribute",
    "ruleId": "@custom-typescript-eslint/jsx-sort-props",
    "severity": 2,
  },
]
`;

exports[`jsx-sort-props rule invalid cases Props in wrong order 1`] = `
[
  {
    "column": 18,
    "endColumn": 23,
    "endLine": 1,
    "line": 1,
    "message": "JSX props should be sorted in alphabetical order. 'a' should come before 'b'.",
    "messageId": "unsortedProps",
    "nodeType": "JSXAttribute",
    "ruleId": "@custom-typescript-eslint/jsx-sort-props",
    "severity": 2,
  },
]
`;

exports[`jsx-sort-props rule invalid cases Similar long names, wrong order 1`] = `
[
  {
    "column": 22,
    "endColumn": 29,
    "endLine": 1,
    "line": 1,
    "message": "JSX props should be sorted in alphabetical order. 'app' should come before 'apple'.",
    "messageId": "unsortedProps",
    "nodeType": "JSXAttribute",
    "ruleId": "@custom-typescript-eslint/jsx-sort-props",
    "severity": 2,
  },
]
`;

exports[`jsx-sort-props rule invalid cases Similar starting letters, wrong order 1`] = `
[
  {
    "column": 19,
    "endColumn": 24,
    "endLine": 1,
    "line": 1,
    "message": "JSX props should be sorted in alphabetical order. 'a' should come before 'aa'.",
    "messageId": "unsortedProps",
    "nodeType": "JSXAttribute",
    "ruleId": "@custom-typescript-eslint/jsx-sort-props",
    "severity": 2,
  },
]
`;

exports[`jsx-sort-props rule invalid cases Spread followed by unsorted props 1`] = `
[
  {
    "column": 29,
    "endColumn": 34,
    "endLine": 1,
    "line": 1,
    "message": "JSX props should be sorted in alphabetical order. 'a' should come before 'b'.",
    "messageId": "unsortedProps",
    "nodeType": "JSXAttribute",
    "ruleId": "@custom-typescript-eslint/jsx-sort-props",
    "severity": 2,
  },
]
`;

exports[`jsx-sort-props rule invalid cases Unsorted with spread between 1`] = `
[
  {
    "column": 29,
    "endColumn": 34,
    "endLine": 1,
    "line": 1,
    "message": "JSX props should be sorted in alphabetical order. 'a' should come before 'b'.",
    "messageId": "unsortedProps",
    "nodeType": "JSXAttribute",
    "ruleId": "@custom-typescript-eslint/jsx-sort-props",
    "severity": 2,
  },
]
`;
